File: sdk/ios/Sources/ContextStoreBuilder.swift
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

import Foundation

/// Fluent builder for a `ContextStore<T>` bound to a Dazzle client namespace.
///
/// ```swift
/// let chat = dazzle.contextStore(
///     name: "chat:42",
///     T: ChatMessage.self,
///     encode: { m in ["role": m.role.rawValue, "text": m.text] },
///     decode: { f in
///         guard let role = f["role"], let text = f["text"] else { return nil }
///         return ChatMessage(role: Role(rawValue: role) ?? .user, text: text)
///     },
///     config: {
///         $0.semanticSearch(dim: 384) { m in embedder.embed(m.text) }
///         $0.timeRange { Int64($0.timestamp.timeIntervalSince1970) }
///         $0.tags { [$0.role.rawValue] }
///     }
/// )
/// ```
public final class ContextStoreBuilder<T: Sendable>: @unchecked Sendable {
    internal var embedder: (@Sendable (T) -> [Float])?
    internal var embeddingDim: Int?
    internal var embedAlgorithm: VectorIndex.Algorithm = .hnsw
    internal var embedMetric: VectorIndex.Metric = .cosine
    internal var timeExtractor: (@Sendable (T) -> Int64)?
    internal var tagsExtractor: (@Sendable (T) -> Set<String>)?

    /// Per-store execution override. When `nil`, inherits from `DazzleConfig`.
    public var execution: ExecutionPolicy?

    /// Declare semantic-search support for this store.
    public func semanticSearch(
        dim: Int,
        algorithm: VectorIndex.Algorithm = .hnsw,
        metric: VectorIndex.Metric = .cosine,
        embed: @escaping @Sendable (T) -> [Float]
    ) {
        precondition(dim > 0, "embedding dim must be positive")
        self.embeddingDim = dim
        self.embedAlgorithm = algorithm
        self.embedMetric = metric
        self.embedder = embed
    }

    /// Declare a timestamp extractor — enables `byTimeRange`.
    public func timeRange(_ extract: @escaping @Sendable (T) -> Int64) {
        timeExtractor = extract
    }

    /// Declare a tag extractor — enables `byTag` / `byTags`.
    public func tags(_ extract: @escaping @Sendable (T) -> Set<String>) {
        tagsExtractor = extract
    }
}

public extension Dazzle {
    /// Build a typed `ContextStore<T>` on top of this Dazzle client.
    ///
    /// - Parameters:
    ///   - name: stable logical identifier (key-prefix).
    ///   - encode: serialize `T` → flat `[String: String]`.
    ///   - decode: reconstruct `T` from stored fields (return `nil` on parse failure).
    ///   - config: optional block to declare index hooks.
    func contextStore<T: Sendable>(
        name: String,
        encode: @escaping @Sendable (T) -> [String: String],
        decode: @escaping @Sendable ([String: String]) -> T?,
        config: (ContextStoreBuilder<T>) -> Void = { _ in }
    ) -> DazzleContextStore<T> {
        let b = ContextStoreBuilder<T>()
        config(b)
        return DazzleContextStore<T>(
            dazzle: self,
            name: name,
            encoder: encode,
            decoder: decode,
            embedder: b.embedder,
            embeddingDim: b.embeddingDim,
            embedAlgorithm: b.embedAlgorithm,
            embedMetric: b.embedMetric,
            timeExtractor: b.timeExtractor,
            tagsExtractor: b.tagsExtractor
        )
    }
}

File: sdk/android/src/main/java/dev/dazzle/sdk/ContextStoreBuilder.kt
// Copyright 2026 Ivan Aliaga <ivan.aliaga@urp.edu.pe>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.

package dev.dazzle.sdk

/**
 * Fluent builder for a [ContextStore] bound to a Dazzle client namespace.
 *
 * Obtain from [Dazzle.contextStore]; call `encode` / `decode` (required)
 * plus any of the optional index hooks, then the store is constructed
 * automatically when the builder lambda returns.
 *
 * ```kotlin
 * val chat = dazzle.contextStore<ChatMessage>("chat:42") {
 *     encode { m -> mapOf("role" to m.role.name, "text" to m.text) }
 *     decode { f -> ChatMessage(Role.valueOf(f["role"]!!), f["text"]!!) }
 *
 *     semanticSearch(dim = 384) { m -> embedder.embed(m.text) }
 *     timeRange { m -> m.timestamp }
 *     tags      { m -> setOf("role:${m.role.name.lowercase()}") }
 *
 *     execution = ExecutionPolicy.parallel  // per-store override; optional
 * }
 * ```
 */
class ContextStoreBuilder<T> internal constructor(
    private val dazzle: Dazzle,
    private val name: String,
) {
    private var encoder: ((T) -> Map<String, String>)? = null
    private var decoder: ((Map<String, String>) -> T)? = null
    private var embedder: ((T) -> FloatArray)? = null
    private var embeddingDim: Int? = null
    private var embedAlgorithm: VectorIndex.Algorithm = VectorIndex.Algorithm.HNSW
    private var embedMetric: VectorIndex.Metric = VectorIndex.Metric.COSINE
    private var timeExtractor: ((T) -> Long)? = null
    private var tagsExtractor: ((T) -> Set<String>)? = null

    /** Per-store execution policy. When null, inherits from `DazzleConfig.execution`. */
    var execution: ExecutionPolicy? = null

    /** Serialize a value into the flat `String → String` shape stored in Valkey.
     *  Every field the caller wants to retrieve later must appear here. */
    fun encode(fn: (T) -> Map<String, String>) { encoder = fn }

    /** Reconstruct a value from its stored fields. Receives the map as written
     *  by [encode] (minus any SDK-reserved fields like `_embedding`). */
    fun decode(fn: (Map<String, String>) -> T) { decoder = fn }

    /**
     * Declare semantic-search support for this store. When set, [put] computes
     * the embedding synchronously and stores it alongside the record, and
     * [ContextStore.semanticSearch] returns hits; when absent, the search
     * methods return empty.
     *
     * @param dim       vector dimensionality. Must match the embedder output.
     * @param algorithm index algorithm (HNSW default; FLAT for small / exact sets).
     * @param metric    distance metric (COSINE default; L2 / IP also supported).
     * @param embed     how to compute the vector from an instance of [T].
     */
    @JvmOverloads
    fun semanticSearch(
        dim: Int,
        algorithm: VectorIndex.Algorithm = VectorIndex.Algorithm.HNSW,
        metric: VectorIndex.Metric = VectorIndex.Metric.COSINE,
        embed: (T) -> FloatArray,
    ) {
        require(dim > 0) { "embedding dim must be positive, got $dim" }
        embeddingDim = dim
        embedAlgorithm = algorithm
        embedMetric = metric
        embedder = embed
    }

    /** Declare a timestamp extractor — enables [ContextStore.byTimeRange].
     *  Timestamps are stored as doubles in a SortedSet; any epoch-based
     *  representation works (millis, seconds, microseconds). */
    fun timeRange(extract: (T) -> Long) { timeExtractor = extract }

    /** Declare a tag extractor — enables [ContextStore.byTag] / [ContextStore.byTags].
     *  Each tag becomes a Valkey Set; SINTER does the intersection natively. */
    fun tags(extract: (T) -> Set<String>) { tagsExtractor = extract }

    internal fun build(): ContextStore<T> {
        val enc = requireNotNull(encoder) {
            "ContextStore '$name' requires encode { ... } — cannot serialize values without it"
        }
        val dec = requireNotNull(decoder) {
            "ContextStore '$name' requires decode { ... } — cannot deserialize values without it"
        }
        return DazzleContextStore(
            dazzle = dazzle,
            name = name,
            encoder = enc,
            decoder = dec,
            embedder = embedder,
            embeddingDim = embeddingDim,
            embedAlgorithm = embedAlgorithm,
            embedMetric = embedMetric,
            timeExtractor = timeExtractor,
            tagsExtractor = tagsExtractor,
            execution = execution,
        )
    }
}

/**
 * Build a typed [ContextStore] on top of this Dazzle client.
 *
 * The store uses a dedicated Valkey key namespace (`cs:<name>:…`) so
 * multiple stores can coexist in the same database without collision.
 *
 * @param name logical identifier — e.g. `"chat:42"`, `"sensors:alpha"`.
 *             Must be stable across app restarts for persistence to work.
 */
fun <T> Dazzle.contextStore(
    name: String,
    build: ContextStoreBuilder<T>.() -> Unit,
): ContextStore<T> = ContextStoreBuilder<T>(this, name).apply(build).build()
